beacon-chain/sync/validate_sync_committee_message.go

```diff
@@ -0,0 +1,310 @@
+// and `subcommittee_index`. In the event of `validator_index` belongs to multiple subnets, as long
+// as one subnet has not been seen, we should let it in.
+func (s *Service) ignoreHasSeenSyncMsg(
+	m *ethpb.SyncCommitteeMessage, committeeIndices []types.CommitteeIndex,
+) validationFn {
+	return func(ctx context.Context) pubsub.ValidationResult {
+		var isValid bool
+		subCommitteeSize := params.BeaconConfig().SyncCommitteeSize / params.BeaconConfig().SyncCommitteeSubnetCount
+		for _, idx := range committeeIndices {
+			subnet := uint64(idx) / subCommitteeSize
+			if s.hasSeenSyncMessageIndexSlot(m.Slot, m.ValidatorIndex, subnet) {
+				isValid = false
+			} else {
+				isValid = true
+			}
+		}
+		if !isValid {
+			return pubsub.ValidationIgnore
+		}
+		return pubsub.ValidationAccept
+	}
+}
+
+func (s *Service) rejectInvalidSignature(m *ethpb.SyncCommitteeMessage) validationFn {
+	return func(ctx context.Context) pubsub.ValidationResult {
+		ctx, span := trace.StartSpan(ctx, "sync.ifInvalidSignature")
+		defer span.End()
+
+		// Ignore the message if it is not possible to retrieve the signing root.
+		d, err := s.cfg.Chain.HeadSyncCommitteeDomain(ctx, m.Slot)
+		if err != nil {
+			traceutil.AnnotateError(span, err)
+			return pubsub.ValidationIgnore
+		}
+		rawBytes := p2ptypes.SSZBytes(m.BlockRoot)
+		sigRoot, err := helpers.ComputeSigningRoot(&rawBytes, d)
+		if err != nil {
+			traceutil.AnnotateError(span, err)
+			return pubsub.ValidationIgnore
+		}
+
+		// Reject for a validator index that is not found.
+		pubKey, err := s.cfg.Chain.HeadValidatorIndexToPublicKey(ctx, m.ValidatorIndex)
+		if err != nil {
+			traceutil.AnnotateError(span, err)
+			return pubsub.ValidationReject
+		}
+
+		// We reject a malformed signature from bytes according to the p2p specification.
+		blsSig, err := bls.SignatureFromBytes(m.Signature)
+		if err != nil {
+			traceutil.AnnotateError(span, err)
+			return pubsub.ValidationReject
+		}
+
+		// Ignore a malformed public key from bytes according to the p2p specification.
+		pKey, err := bls.PublicKeyFromBytes(pubKey[:])
+		if err != nil {
+			traceutil.AnnotateError(span, err)
+			return pubsub.ValidationIgnore
+		}
+
+		verified := blsSig.Verify(pKey, sigRoot[:])
+		if !verified {
+			return pubsub.ValidationReject
+		}
+		return pubsub.ValidationAccept
+	}
+}
+
+func rejectNilTopic(msg *pubsub.Message) validationFn {
+	return func(ctx context.Context) pubsub.ValidationResult {
+		if msg.Topic == nil {
+			return pubsub.ValidationReject
+		}
+		return pubsub.ValidationAccept
+	}
+}
+
+func ignoreEmptyCommittee(indices []types.CommitteeIndex) validationFn {
+	return func(ctx context.Context) pubsub.ValidationResult {
+		if len(indices) == 0 {
+			return pubsub.ValidationIgnore
+		}
+		return pubsub.ValidationAccept
+	}
+}
+
+func withValidationPipeline(ctx context.Context, fns ...validationFn) pubsub.ValidationResult {
```

Comment:
The functional options pattern in golang is often as a builder using method names like `validator.WithIgnoreEmptyCommittee().WithRejectNilTopic`. So this is totally a nitpick, but `withValidationPipeline` throws me off every time I see it. A name like `validatePipeline`, `evaluatePipeline`, `getPipelineResult` would be more clear, to me.